File: src/main/java/io/github/liyze09/arms/network/Connection.java
package io.github.liyze09.arms.network;

import io.netty.channel.ChannelFuture;
import io.netty.channel.ChannelHandlerContext;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

public final class Connection {
    @Contract("_ -> new")
    public static @NotNull Connection addConnection(ChannelHandlerContext ctx) {
        return new Connection(
                Objects.requireNonNull(ctx)
        );
    }

    public static @NotNull Connection getInstance(ChannelHandlerContext session) {
        Objects.requireNonNull(session);
        var connection = connections.get(session);
        if (connection == null) {
            connection = addConnection(session);
        }
        return connection;
    }

    public static ChannelFuture disconnect(ChannelHandlerContext session) {
        connections.remove(session);
        return session.close();
    }

    private static final Map<ChannelHandlerContext, Connection> connections = new ConcurrentHashMap<>();

    private Connection(ChannelHandlerContext ctx) {
        this.ctx = ctx;
        connections.put(ctx, this);
    }

    public Status getStatus() {
        return this.status;
    }

    public synchronized void updateStatus(Status status) {
        Objects.requireNonNull(status);
        this.status = status;
    }

    public MainHand getMainHand() {
        if(mainHand == null) {
            throw new IllegalStateException();
        }
        return mainHand;
    }

    public void updateMainHand(MainHand mainHand) {
        Objects.requireNonNull(mainHand);
        this.mainHand = mainHand;
    }

    public boolean isAllowServerListings() {
        return allowServerListings;
    }

    public void updateAllowServerListings(boolean allowServerListings) {
        this.allowServerListings = allowServerListings;
    }

    public DisplayedSkinParts getDisplayedSkinParts() {
        if (displayedSkinParts == null) {
            throw new IllegalStateException();
        }
        return displayedSkinParts;
    }

    public void updateDisplayedSkinParts(DisplayedSkinParts displayedSkinParts) {
        Objects.requireNonNull(displayedSkinParts);
        this.displayedSkinParts = displayedSkinParts;
    }

    public boolean isChatColors() {
        return chatColors;
    }

    public void updateChatColors(boolean chatColors) {
        this.chatColors = chatColors;
    }

    public String getName() {
        if (displayedSkinParts == null) {
            throw new IllegalStateException();
        }
        return username;
    }

    public String getLocale() {
        if (locale == null) {
            throw new IllegalStateException();
        }
        return locale;
    }

    public void updateLocale(String locale) {
        Objects.requireNonNull(locale);
        this.locale = locale;
    }

    public byte getViewDistance() {
        if (viewDistance == -1) {
            throw new IllegalStateException();
        }
        return viewDistance;
    }

    public void updateViewDistance(byte viewDistance) {
        if (viewDistance < 0) throw new IllegalArgumentException();
        this.viewDistance = viewDistance;
    }

    public ChatMode getChatMode() {
        return chatMode;
    }

    public void updateChatMode(ChatMode chatMode) {
        Objects.requireNonNull(chatMode);
        this.chatMode = chatMode;
    }

    public int getProtocolVersion() {
        if (protocolVersion == -1) {
            throw new IllegalStateException();
        }
        return protocolVersion;
    }

    public void setProtocolVersion(int protocolVersion) {
        if (protocolVersion < 0) throw new IllegalArgumentException();
        if (this.protocolVersion != -1) throw new IllegalStateException();
        this.protocolVersion = protocolVersion;
    }

    public void setName(String username) {
        Objects.requireNonNull(username);
        if (this.username != null) throw new IllegalStateException();
        this.username = username;
    }

    public UUID getUUID() {
        if (uuid == null) {
            throw new IllegalStateException();
        }
        return uuid;
    }

    public void setUUID(UUID uuid) {
        Objects.requireNonNull(uuid);
        if (this.uuid != null) throw new IllegalStateException();
        this.uuid = uuid;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Connection that)) return false;
        return Objects.equals(ctx, that.ctx);
    }

    @Override
    public int hashCode() {
        return ctx.hashCode();
    }

    @Override
    public String toString() {
        return "Connection{" +
                "context=" + ctx.name() +
                ", username='" + username +
                ", uuid=" + uuid.toString() +
                ", protocolVersion=" + protocolVersion +
                ", status=" + status +
                ", locale='" + locale +
                ", viewDistance=" + viewDistance +
                ", chatMode=" + chatMode +
                ", chatColors=" + chatColors +
                ", displayedSkinParts=" + displayedSkinParts +
                ", allowServerListings=" + allowServerListings +
                ", mainHand=" + mainHand +
                '}';
    }

    public enum Status {
        HANDSHAKE,
        LOGIN,
        PLAY,
        CONFIGURATION,
        STATUS
    }

    public record UUID(long a, long b) {
        @Override
        public @NotNull String toString() {
            return String.format("%016x%016x", a, b);
        }
    }

    public enum ChatMode {
        ENABLED,
        COMMANDS_ONLY,
        HIDDEN
    }

    public record DisplayedSkinParts(boolean cape,
                                     boolean jacket,
                                     boolean leftSleeve,
                                     boolean rightSleeve,
                                     boolean leftPantsLeg,
                                     boolean rightPantsLeg,
                                     boolean hat
    ) {
    }

    public enum MainHand {
        LEFT,
        RIGHT
    }

    public final ChannelHandlerContext ctx;
    private String username = null;
    private UUID uuid = null;
    private int protocolVersion = -1;
    private Status status = Status.HANDSHAKE;
    private String locale = null;
    private byte viewDistance = -1;
    private ChatMode chatMode = null;
    private boolean chatColors = false;
    private DisplayedSkinParts displayedSkinParts = null;
    private MainHand mainHand = null;
    private boolean allowServerListings = true;
}

File: src/main/kotlin/io/github/liyze09/arms/network/Connection.kt
package io.github.liyze09.arms.network

import io.github.liyze09.arms.network.NettyInitialize.LOGGER
import io.netty.channel.ChannelFuture
import io.netty.channel.ChannelHandlerContext
import org.jetbrains.annotations.Contract
import java.io.IOException
import java.util.*
import java.util.concurrent.ConcurrentHashMap

class Connection private constructor(@JvmField val ctx: ChannelHandlerContext) {
    @Synchronized
    fun updateStatus(status: Status) {
        this.status = status
    }

    fun getMainHand(): MainHand {
        checkNotNull(mainHand)
        return mainHand as MainHand
    }

    fun updateMainHand(mainHand: MainHand) {
        Objects.requireNonNull(mainHand)
        this.mainHand = mainHand
    }

    fun updateAllowServerListings(allowServerListings: Boolean) {
        this.isAllowServerListings = allowServerListings
    }

    fun getDisplayedSkinParts(): DisplayedSkinParts {
        checkNotNull(displayedSkinParts)
        return displayedSkinParts as DisplayedSkinParts
    }

    fun updateDisplayedSkinParts(displayedSkinParts: DisplayedSkinParts) {
        Objects.requireNonNull(displayedSkinParts)
        this.displayedSkinParts = displayedSkinParts
    }

    fun updateChatColors(chatColors: Boolean) {
        this.isChatColors = chatColors
    }

    fun getUsername(): String {
        checkNotNull(username)
        return username as String
    }

    fun setUsername(username: String) {
        check(this.username == null)
        this.username = username
    }

    fun getLocale(): String {
        checkNotNull(locale)
        return locale as String
    }

    fun updateLocale(locale: String) {
        this.locale = locale
    }

    fun getViewDistance(): Byte {
        check(viewDistance.toInt() != -1)
        return viewDistance
    }

    fun updateViewDistance(viewDistance: Byte) {
        require(viewDistance >= 0)
        this.viewDistance = viewDistance
    }

    fun updateChatMode(chatMode: ChatMode) {
        this.chatMode = chatMode
    }

    fun getProtocolVersion(): Int {
        check(protocolVersion != -1)
        return protocolVersion
    }

    fun setProtocolVersion(protocolVersion: Int) {
        require(protocolVersion >= 0)
        check(this.protocolVersion == -1)
        this.protocolVersion = protocolVersion
    }

    fun getUUID(): UUID {
        checkNotNull(uuid)
        return uuid as UUID
    }

    fun setUUID(uuid: UUID) {
        check(this.uuid == null)
        this.uuid = uuid
    }

    fun <T> sendPacket(msg: T, encoder: io.github.liyze09.arms.network.packet.ClientBoundPacketEncoder<T>): Connection {
        try {
            val packet = encoder.encode(msg, this)
            LOGGER.debug("To {}: {}", this.ctx.name(), packet)
            this.ctx.writeAndFlush(packet)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        return this
    }

    @Volatile
    var teleporting = -1

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Connection) return false
        return ctx == other.ctx
    }

    override fun hashCode(): Int = ctx.hashCode()


    override fun toString(): String {
        return "Connection{" +
                "context=" + ctx.name() +
                ", username='" + username +
                ", uuid=" + uuid.toString() +
                ", protocolVersion=" + protocolVersion +
                ", status=" + status +
                ", locale='" + locale +
                ", viewDistance=" + viewDistance +
                ", chatMode=" + chatMode +
                ", chatColors=" + isChatColors +
                ", displayedSkinParts=" + displayedSkinParts +
                ", allowServerListings=" + isAllowServerListings +
                ", mainHand=" + mainHand +
                '}'
    }

    fun getStatus(): Status = status

    enum class Status {
        HANDSHAKE,
        LOGIN,
        PLAY,
        CONFIGURATION,
        STATUS
    }

    @JvmRecord
    data class UUID(val a: Long, val b: Long) {
        override fun toString(): String = String.format("%016x%016x", a, b)
    }

    enum class ChatMode {
        ENABLED,
        COMMANDS_ONLY,
        HIDDEN
    }

    @JvmRecord
    data class DisplayedSkinParts(
        val cape: Boolean,
        val jacket: Boolean,
        val leftSleeve: Boolean,
        val rightSleeve: Boolean,
        val leftPantsLeg: Boolean,
        val rightPantsLeg: Boolean,
        val hat: Boolean
    )

    enum class MainHand {
        LEFT,
        RIGHT
    }

    private var username: String? = null
    private var uuid: UUID? = null
    internal var protocolVersion = -1
    private var status: Status = Status.HANDSHAKE
    private var locale: String? = null
    private var viewDistance: Byte = -1
    private var chatMode: ChatMode? = null
    private var isChatColors: Boolean = false
    private var displayedSkinParts: DisplayedSkinParts? = null
    private var mainHand: MainHand? = null
    private var isAllowServerListings: Boolean = true

    init {
        connections[ctx] = this
    }

    companion object {
        @Contract("_ -> new")
        internal fun addConnection(ctx: ChannelHandlerContext): Connection = Connection(ctx)

        @JvmStatic
        fun getInstance(session: ChannelHandlerContext): Connection {
            Objects.requireNonNull(session)
            var connection = connections[session]
            if (connection == null) {
                connection = addConnection(session)
            }
            return connection
        }

        fun disconnect(session: ChannelHandlerContext): ChannelFuture {
            connections.remove(session)
            return session.close()
        }

        private val connections: MutableMap<ChannelHandlerContext, Connection> = ConcurrentHashMap()
    }
}
